feat(meta): URL field schema and client-side validation (mocked)

cristianoliveira opened the pull request:
This commit modifies the user metadata client to mock URL field schemas and removes client-side URL validation in the UI.

Detailed Changes:
 - Implementation:
   - Modified MetaClient.js to mock URL field schemas by replacing properties containing '-url' with a mock schema.
   - Changed URL.jsx to remove client-side URL validation logic and state management.
   - Fixed a missing newline at the end of DateTime.js.

Plus mocks setup for vitest so we can mock global pydio and other libraries

### DEMO

https://github.com/user-attachments/assets/f3548955-00a4-40b0-91b7-5162e429fb87

cristianoliveira committed:
feat(meta): URL field schema and client-side validation (mocked)

This commit modifies the user metadata client to mock URL field schemas and removes client-side URL validation in the UI.

Detailed Changes:
 - Implementation:
   - Modified MetaClient.js to mock URL field schemas by replacing properties containing '-url' with a mock schema.
   - Changed URL.jsx to remove client-side URL validation logic and state management.
   - Fixed a missing newline at the end of DateTime.js.

feat: use mocked json schema

(cherry picked from commit fff69aa9a02d5ab5740ca5be2f0181c85fefbc48)

cristianoliveira committed:
chore: delete dead code

(cherry picked from commit 7a6ce8401a4631c676897e629fc02403b2a90011)

cristianoliveira commented on frontend/assets/meta.user/res/js/UserMetaPanel.js:
```diff
@@ -248,6 +250,9 @@ export default class UserMetaPanel extends React.Component{
                     case 'boolean':
                         field = <BooleanForm {...baseProps}/>
                         break
+                    case 'url': // FIXME: will it be url only?
```
here will probably be `external_url`?

osalama7 replied:
maybe also url for admin side the look up object is
```
Metadata.MetaTypes = {
    "string": "Text",
    "textarea": "Long Text",
    "integer": "Number",
    "boolean": "Boolean",
    "date": "Date",
    "choice": "Selection",
    "tags": "Extensible Tags",
    "stars_rate": "Stars Rating",
    "css_label": "Color Labels",
    "json": "JSON",
    "url": "External URL", // can be used.
};
```

cristianoliveira replied:
Perfect, updated the other PR https://github.com/pydio/cells/pull/745/files#diff-094996decae6adfb3523ee58a05588e718f8d2ad1e66a20359cf73e92099ea98R94

cristianoliveira commented on frontend/assets/meta.user/res/js/fields/URL.jsx:
```diff
@@ -0,0 +1,162 @@
+                    <a
+                        href={previewUrl}
+                        target="_blank"
+                        aria-label={`Open ${value} in a new tab`}
```
a11y <3

cristianoliveira replied:
TODO: needs translations most likely

cristianoliveira committed:
chore: ignore mocks when building

(cherry picked from commit 2eb50bc828d1740298d1753c34b3bd8a722fa477)

cristianoliveira committed:
chore: build for prod

cristianoliveira committed:
test: add title to URL links and add @testing-library/jest-dom matchers

This commit adds a title attribute to URL links for better accessibility and updates testing dependencies and configuration.

Detailed Changes:
 - Implementation:
   - Added title attribute to URL links in URL.jsx to show the URL on hover.
 - Tests:
   - Updated URL.spec.jsx to use `toBeInTheDocument()` matcher and removed unused imports.
   - Added vitest-setup.js to import jest-dom matchers.
   - Updated vitest.config.js to include the setup file.
 - Dependencies:
   - Updated package.json to reorder devDependencies and upgrade @testing-library/jest-dom.
   - Updated pnpm-lock.yaml with new dependency versions and changed cells-sdk source.
 - Development Experience:
   - Added vitest-setup.js for custom test setup.

(cherry picked from commit 57c5cd75aa2781d429e732ff59af7d3a0156af89)

cristianoliveira committed:
feat: url adds 300ms debounce before validate

(cherry picked from commit c89542807ad83a8f33c2965db9df88237ec5ccfe)

cristianoliveira committed:
chore: build for prod

cristianoliveira committed:
chore(devx): add JSDoc types and fix URL field tests

This commit adds JSDoc comments to the URL field components and updates the test suite to use fake timers for debounced input.

Detailed Changes:
 - Implementation:
   - Added JSDoc comments to URLIcon, URLLinkIcon, URLFieldBase, URLField, URLFormBase, and URLForm components.
   - Changed URLLinkIcon to accept a displayText prop and use it for aria-label.
 - Tests:
   - Updated URL.spec.jsx to import waitFor from testing library.
   - Removed test for automatic https:// prefix addition.
   - Changed test for updateValue to use fake timers and wait for debounced call.

(cherry picked from commit ceeb6b62d32028e233fb650027285d1885745b46)

cristianoliveira commented on frontend/assets/vitest-setup.js:
```diff
@@ -0,0 +1,2 @@
+// In your own vitest-setup.js (or any other name)
```
```suggestion
```

cristianoliveira committed:
chore: remove comment

cristianoliveira committed:
chore: remove dead code

(cherry picked from commit 4f19dd50091771370bcb573216b1e859a3804cdf)

osalama7 commented on frontend/assets/meta.user/res/js/fields/URL.jsx:
```diff
@@ -0,0 +1,220 @@
+ *  @returns {React.ReactElement}
+ */
+const URLLinkIcon = ({ fontSize, url, displayText, children }) => {
+    const hasMinimalValidity = /^https?:\/\//i.test(url);
```
We can allow http:// urls as well at some point

cristianoliveira replied:
This already cover the http, the s is optional in this regex :)

osalama7 commented on frontend/assets/meta.user/res/js/fields/URL.jsx:
```diff
@@ -0,0 +1,220 @@
+ * @returns {React.ReactElement}
+ */
+const URLFieldBase = ({ getRealValue }) => {
+    const value = getRealValue();
```
Nitpick May be simplified 
```suggestion
    const url = getRealValue();
```

cristianoliveira replied:
Changed since now I use a sanitized URL

osalama7 reviewed (approved):
lgtm, added vitest as well 👍

cristianoliveira committed:
fix(meta): sanitize URL fields and update link color

This commit adds URL sanitization to prevent XSS attacks and changes the link color from primary to secondary.

Detailed Changes:
 - Implementation:
   - Added @braintree/sanitize-url dependency to package.json and pnpm-lock.yaml.
   - Changed URL validation to use sanitizeUrl function in URLLinkIcon and URLFieldBase components.
   - Changed link color from primary to secondary in URLLinkIcon.
   - Added sanitizeUrl import and removed URLIcon from children in URLLinkIcon.
   - Updated URLFormBase to accept rest props and changed hintText default.
 - Dependencies:
   - Added @braintree/sanitize-url version 7.1.1 to meta.user dependencies.

(cherry picked from commit 46e93e086a5ffcafdb0640194b6600e74fd674ce)

cristianoliveira committed:
fix(meta): sanitize URL fields and update link color

This commit changes URL field validation to add http scheme when missing and removes debounce.

Detailed Changes:
 - Implementation:
   - Changed URL.jsx to add ensureHttpScheme function that prepends https:// to URLs without a scheme.
   - Changed URL.jsx to remove debounced validation and replace with onBlur validation.
   - Changed URL.jsx to sanitize URLs after scheme normalization.
 - Tests:
   - Changed URL.spec.jsx to update mock and add tests for http scheme injection and blur behavior.

(cherry picked from commit 70f37faebcc78deed01c76b4c004eb582a6c3c24)

cristianoliveira committed:
feat(ui): change URL field icon color and positioning

Changed the URL field icon color to inherit from parent and adjusted its vertical position.

Detailed Changes:
 - Implementation:
   - Changed URLIcon component to inherit color instead of using a CSS variable.
   - Changed URLLinkIcon component to inherit color instead of using a CSS variable.
   - Changed URLFormBase component to position the link icon at the bottom instead of the top.

(cherry picked from commit f21caa3adedd942031cf0cbf0097f7f0c1a111c6)

cristianoliveira committed:
feat(meta): remove mock schema and definition from MetaClient (#746)

Removed mock JSON schema and definition handling for namespaces containing '-url' in MetaClient.

Detailed Changes:
 - Implementation:
   - Removed mock schema transformation in getNamespaceSchema method.
   - Removed mock schema and definition assignment for '-url' namespaces in getNamespaceConfigs method.

(cherry picked from commit 81be58ae6d67bc3209477ccd316f74970961261f)
